lib/fastlane/plugin/test_center/helper/html_test_report.rb

```diff
@@ -181,11 +181,20 @@ def duplicate_testcases?
         end
 
         def remove_duplicate_testcases
-          nonuniq_testcases = testcases
-          uniq_testcases = nonuniq_testcases.uniq { |tc| tc.title }
-          (nonuniq_testcases - uniq_testcases).each do |tc|
+          # get a list of all the testcases in the report's testsuite
+          # and order them according to their passing status
+          # that way, `uniq`, which traverses the array in order, keeping the first
+          # will have the passing case of a list of duplicated testcases
+          ordered_nonuniq_testcases = testcases.sort_by { |tc| tc.passing? ? 0 : 1 }
+          # for example, if `ordered_nonuniq_testcases` is
+          # `['a(passing)', 'b(passing)', 'c(passing)', 'dup1(passing)', 'dup2(passing)', 'dup2(failing)', 'dup2(failing)' ]`
+          # then `uniq_testcases` will be
+          # `['a(passing)', 'b(passing)', 'c(passing)', 'dup1(passing)', 'dup2(passing)']`
+          uniq_testcases = ordered_nonuniq_testcases.uniq { |tc| tc.title }
+          (ordered_nonuniq_testcases - uniq_testcases).each do |tc|
+            # here, we would be deleting ['dup1(failing)', 'dup2(failing)']
```

Comment:
`sort_by` is not a stable sort. I assume the input HTML file is sorted in the same order tests were run (which is typically alphabetically, unless the user has requested randomized tests), right? It seems valuable to preserve the existing ordering. This suggests that you may have to reimplement `uniq` yourself to give you control over which one is preserved.

I'm no Ruby expert so maybe there's something I'm missing, but this is what I came up with, given that `Hash` apparently preserves insertion order of keys:

```suggestion
          uniq_testcases = testcases.each_with_object({}) { |tc,keep|
              title = tc.title
              old = keep[title]
              if !old.nil? && tc.passing? && !old.passing?
                  keep.delete(title)
              end
              keep[title] = tc unless keep.key?(title)
          }.values
          uniq_testcases.each do |tc|
```

This uniques all testcases, keeping the first passing instance of each case, or the first failing instance if none passed, and it preserves order. This does mean if tests are run alphabetically the results will be non-alphabetical if a test failed and then passed on retry, because the retry will happen later. I can't decide if this is appropriate or if we should move the passing retry up to take the place of the original failure. That behavior can be done with

```suggestion
          uniq_testcases = testcases.each_with_object({}) { |tc,keep|
              title = tc.title
              old = keep[title]
              keep[title] = tc if old.nil? || (tc.passing? && !old.passing?)
          }.values
          uniq_testcases.each do |tc|
```

Reply:
The reason that there are duplicated testcases is because the failing tests can appear multiple times until they pass. So, really, the implementation should grab the last of the duplicate tests (regardless of whether or not the test passed: it was the most recent result).

There shouldn't be any reason for a failing test to appear after a passing test as multi_scan should be removing it from the list of tests to test ^1.

I don't think I have to reimplement `uniq`: my intuition says that I can simply reverse the order of the value returned from `testcases`. For example, say it is:

```
[
  'a(passing)', 
  'b(passing)',
  'c(passing)',
  'dup1(failing)',
  'dup1(passing)',
  'dup2(failing)',
  'dup2(passing)' 
]
```

reversing that will give me:
```
[
  'dup2(passing)',
  'dup2(failing)',
  'dup1(passing)',
  'dup1(failing)',
  'c(passing)',
  'b(passing)',
  'a(passing)' 
]
```

uniq will give me:
```
[
  'dup2(passing)',
  'dup1(passing)',
  'c(passing)',
  'b(passing)',
  'a(passing)' 
]
```

and the difference will be:
```
[
  'dup2(failing)',
  'dup1(failing)',
]
```

Which will then be removed from the report. Thoughts?

---
[1] there is a feature request to allow people to control which tests they run for new re-runs so, it is possible for a test to fail after a pass if the user explicitly requested that test to re-run.

Reply:
That's a good point about failing tests should always be after the equivalent passing. This is assuming the html report lists tests in the order they were executed, but I assume that's a safe assumption.

Also I kind of forgot the point of uniquing test cases was to then calculate what to remove, as opposed to what to keep, meaning the distinction I made about ordering between my two suggestions is meaningless.

In any case, given the reasonable assumption about test run relative ordering, uniq'ing the reversed list to grab the last run of each test seems like a reasonable approach.